dss/util/es.py

```diff
@@ -34,36 +40,99 @@ def __call__(self, r):
         return r
 
 
+class ElasticsearchServer:
+    def __init__(self, startup_timeout_seconds: int=60) -> None:
+        elasticsearch_binary = os.getenv("DSS_TEST_ES_PATH")
+        self.tempdir = tempfile.TemporaryDirectory()
+        while True:
+            port = random.randint(1024, 65535)
+            transport_port = random.randint(1024, 65535)
+
+            # try to check if the randomly assigned ports already have a running service.
+            conflict = False
+            for port_to_test in port, transport_port:
+                try:
+                    sock = socket.create_connection(("localhost", port_to_test), 1)
+                    sock.close()
+                    # there's something there already.
+                    conflict = True
+                    break
+                except (ConnectionRefusedError, socket.timeout):
+                    # ok this is an open port.
+                    pass
+            if conflict:
+                continue
+
+            proc = subprocess.Popen(
+                [
+                    elasticsearch_binary,
+                    "-E", f"http.port={port}",
+                    "-E", f"transport.tcp.port={transport_port}",
+                    "-E", f"path.data={self.tempdir.name}",
+                ],
+            )
```

Comment:
Here is a cute way to make this a bit simpler:

`elasticsearch -E http.port=0 -E transport.tcp.port=0 -E path.data=$(mktemp -d)`

0 means let the kernel allocate the port. But I couldn't find a nice way to read back the port number, short of parsing the log.